### ISSUE TITLE
Save server details for non-favorites

### ISSUE THREAD

**Issue by trossyvormes** [mumble-voip/mumble#3713]
Hi,

I plan on adding an option to save server details (hostname, port, username, password) in the database for non-favorite servers. This is useful for servers that don't support certificate authentication and which users connect to primarily via a `mumble://` URI.

Are there any strong objections to this? If not, I will build it and submit a PR soon.

**Comment by davidebeatrici** [comment]
Hi,

Sounds good, as long as the saved servers can be managed.

**Comment by trossyvormes** [comment]
I'm thinking another category under the server connect dialog called 'Saved' or something (to not clutter the 'Favorites' category). The names of the servers would default to hostname:port. The user could move them into 'Favorites' if they want. How does that sound?

**Comment by davidebeatrici** [comment]
Excellent! Perhaps "Recent" is a suitable name for the category?

**Comment by trossyvormes** [comment]
That sounds good to me. Maybe this calls for a prototype category implementation? We have three options here:

1. Add a string `category` field to the `servers` table
2. Add a `categories` table and a `category` FK to servers
3. Add an `is_recent` field to `servers`

Let me know what you prefer. If we go either of the category routes, we could make them hidden for this feature, and in the future add functionality to manage them.
-------- Original Message --------

**Comment by davidebeatrici** [comment]
Do we want favorite servers to show up in the recent list or would it be redundant?

**Comment by trossyvormes** [comment]
Either way works I think from a UX perspective. I think it's easier to have them not show up since all three methods I mentioned use the same table as favorites (if a server was marked recent, through is_recent or category, it wouldn't show up in favorites.) We can figure out some other ways if you think favorites showing in recents is important

**Comment by davidebeatrici** [comment]
Now that I think about it, splitting the dialog in multiple tabs would probably make sense:

1. Favorites
2. Recent
3. Internet

In that case not seeing a favorite server in the Recent tab would probably be confusing though.

I think that the second option would be best, because it also allows us to add more categories in future, perhaps user-defined.

**Comment by trossyvormes** [comment]
>Now that I think about it, splitting the dialog in multiple tabs would probably make sense:

This makes sense to me. By tabs you mean root tree elements, yeah?

> In that case not seeing a favorite server in the Recent tab would probably be confusing though.

We can have servers show in both recents and categories, however that would involve both an is_recent (or date_connected) column as well as a categories column on the servers table. I'm okay with that if you are.

> I think that the second option would be best, because it also allows us to add more categories in future, perhaps user-defined

I think the second option makes sense as well.

> Is it possible to define multiple categories per server without creating a table for each one?

Yes, we just create a [many-to-many](https://en.wikipedia.org/wiki/Many-to-many_(data_model)) relationship between servers and categories .

**Comment by davidebeatrici** [comment]
> This makes sense to me. By tabs you mean root tree elements, yeah?

![](https://user-images.githubusercontent.com/5897523/60472143-e8095100-9c67-11e9-990a-d3b001c4fc6b.png)

> Yes, we just create a [many-to-many](https://en.wikipedia.org/wiki/Many-to-many_(data_model)) relationship between servers and categories .

Excellent, I think that would be the best solution.
It also allows us to show a server in both the recent and favorites lists without any additional fields.

**Comment by trossyvormes** [comment]
>It also allows us to show a server in both the recent and favorites lists without any additional fields.

Yeah, that's the other (probably better) way we could do that. The only disadvantage to it over `date_connected` is that the user has to manually remove it from favorites.

For the dialog layout, I was thinking we could re-use the [tree layout we have currently](https://quad.pe/e/kt3vf9MwNw.png) (right now it just has "Favorite" and "Public Internet"), but if you want to go with tabs that works for me too.

**Comment by davidebeatrici** [comment]
> Yeah, that's the other (probably better) way we could do that. The only disadvantage to it over `date_connected` is that the user has to manually remove it from favorites.

You mean from the recent list?

> For the dialog layout, I was thinking we could re-use the tree layout we have currently (right now it just has "Favorite" and "Public Internet"), but if you want to go with tabs that works for me too.

Indeed, we can re-use it. I was mostly concerned about the height of the dialog, but the sections can be collapsed.

**Comment by trossyvormes** [comment]
>You mean from the recent list?

Whoops, yes. It's not a huge deal anyway - we could automatically uncheck the "Recent" category when the edit dialog is opened for a server

>Indeed, we can re-use it. I was mostly concerned about the height of the dialog, but the sections can be collapsed.

I think re-using it is the best way to go, if a user has many categories then tabs won't work extremely well (though I don't expect many users to have a lot of categories)

**Comment by davidebeatrici** [comment]
I agree. Also, less changes are needed.

**Comment by Kissaki** [comment]
> I plan on adding an option to save server details (hostname, port, username, password) in the database for non-favorite servers. This is useful for servers that don't support certificate authentication and which users connect to primarily via a `mumble://` URI.

I do not understand. Can those not be saved as favourites?

If users connect to them primarily via a mumble URI, your use case would then be connecting to the server via the mumble URI only the first time?

I like the idea of recent servers.

Not so sure about categories, or implementing the possibility for categories when we have no UI for them. We can do that later when we actually implement it. We don't gain anything by introducing categories in the DB when all we want to do is display most recent servers.

I don’t think I’ve seen any requests for categories of favourite servers, so I don’t think we need it. And to keep the UI simple I don’t think we should add it in that case.

Personally I think the recent servers is probably a preferable default over favourites. But that is debatable. Maybe it should be configurable. Especially if it becomes tabs where it is a harder one or the other instead of collapsible lists in the same tab/window.

I’m not sure about using tabs. Tabs probably make sense when there are a lot of/numerous servers in each tab. But is that even a common use case? Given that we have not seen feature request or talk about managing a long list of favourite servers I don’t think so.

**Comment by trossyvormes** [comment]
> I do not understand. Can those not be saved as favourites?
> 
> If users connect to them primarily via a mumble URI, your use case would then be connecting to the server via the mumble URI only the first time?

The primary use case is so that people can connect to servers using URLs without having to re-enter passwords.

> Not so sure about categories, or implementing the possibility for categories when we have no UI for them. We can do that later when we actually implement it. We don't gain anything by introducing categories in the DB when all we want to do is display most recent servers.

I disagree. Introducing categories into the DB right now is a good way to go because it future-proofs us if/when we _do_ eventually want to introduce categories. Introducing categories is not a major database change, and the other options, while simpler, will cause pain in the future when migrating to categories.

> I don’t think I’ve seen any requests for categories of favourite servers, so I don’t think we need it. And to keep the UI simple I don’t think we should add it in that case.

We don't need to actually introduce categories into the UI in a major way - just add another entry to the collapsible list and a checkbox on the server edit dialog.

> Personally I think the recent servers is probably a preferable default over favourites. But that is debatable. Maybe it should be configurable. Especially if it becomes tabs where it is a harder one or the other instead of collapsible lists in the same tab/window.
> 
> I’m not sure about using tabs. Tabs probably make sense when there are a lot of/numerous servers in each tab. But is that even a common use case? Given that we have not seen feature request or talk about managing a long list of favourite servers I don’t think so.

Yeah, we agreed above on using the existing collapsible list over tabs.
